druid/src/main/scala/akka/stream/alpakka/druid/internal/TranquilizerStage.scala

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.alpakka.druid.internal
+
+import java.util.concurrent.atomic.AtomicInteger
+
+import akka.stream.alpakka.druid.TranquilizerSettings
+import akka.stream.{ Attributes, FlowShape, Inlet, Outlet }
+import akka.stream.stage._
+import com.metamx.tranquility.druid.DruidBeams
+import com.metamx.tranquility.tranquilizer.{ MessageDroppedException, Tranquilizer }
+import com.twitter.util.{ Return, Throw }
+import org.slf4j.LoggerFactory
+
+import scala.concurrent.{ Future, Promise }
+import scala.util.{ Failure, Success, Try }
+
+class TranquilizerStage[A](settings: TranquilizerSettings[A]) extends GraphStage[FlowShape[A, Future[A]]] {
+
+  private val in = Inlet[A]("messages")
+  private val out = Outlet[Future[A]]("result")
+
+  override val shape = FlowShape(in, out)
+
+  @scala.throws[Exception](classOf[Exception])
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = {
+
+    val druidSender: Tranquilizer[A] = DruidBeams
+      .fromConfig(settings.dataSourceConfig, settings.timestamper, settings.objectWriter)
+      .partitioner(settings.partitioner)
+      .buildTranquilizer(settings.dataSourceConfig.tranquilizerBuilder())
+
+    new GraphStageLogic(shape) with StageLogging {
+
+      val awaitingConfirmation = new AtomicInteger(0)
+      @volatile var inIsClosed = false
+      var completionState: Option[Try[Unit]] = None
+
+      override def preStart(): Unit = {
+        super.preStart()
+
+        druidSender.start()
+
+      }
+
+      setHandler(out,
+        new OutHandler {
+        override def onPull() =
+          tryPull(in)
+      })
+
+      setHandler(in,
+        new InHandler {
+        override def onPush() = {
+          val msg = grab(in)
+          val r = Promise[A]
+
+          druidSender.send(msg) respond {
+            case Return(_) =>
+              if (log.isDebugEnabled)
+                log.debug(s"Sent message: $msg")
+              r.success(msg)
+            case Throw(e: MessageDroppedException) =>
+              if (log.isDebugEnabled)
+                log.debug(s"Dropped message: $msg", e)
+              r.failure(e)
+            case Throw(e) =>
+              log.error("Failed to send message: $values", e)
+              r.failure(e)
+          } ensure {
+            if (awaitingConfirmation.decrementAndGet() == 0 && inIsClosed)
+              checkForCompletionCB.invoke(())
+          }
+
+          awaitingConfirmation.incrementAndGet()
+          push(out, r.future)
+        }
+
+        override def onUpstreamFinish() = {
+          inIsClosed = true
+          completionState = Some(Success(()))
+          checkForCompletion()
```

Comment:
alternatively, move this to postStop. Then you will not forget to implement `onDownstreamFinish` ;)